File: mRemoteV1/UI/UI.Window.PortScan.vb
Imports System.Net
Imports mRemote3G.App
Imports mRemote3G.Connection.Protocol
Imports mRemote3G.Messages
Imports mRemote3G.Tools
Imports mRemote3G.Tools.PortScan
Imports WeifenLuo.WinFormsUI.Docking

Namespace UI

    Namespace Window
        Public Class PortScan
            Inherits Base

#Region "Constructors"

            Public Sub New(panel As DockContent, import As Boolean)
                InitializeComponent()

                WindowType = Type.PortScan
                DockPnl = panel
                _import = import
            End Sub

#End Region

#Region "Private Properties"

            Private ReadOnly Property IpsValid As Boolean
                Get
                    If String.IsNullOrEmpty(ipStart.Octet1) Then Return False
                    If String.IsNullOrEmpty(ipStart.Octet2) Then Return False
                    If String.IsNullOrEmpty(ipStart.Octet3) Then Return False
                    If String.IsNullOrEmpty(ipStart.Octet4) Then Return False

                    If String.IsNullOrEmpty(ipEnd.Octet1) Then Return False
                    If String.IsNullOrEmpty(ipEnd.Octet2) Then Return False
                    If String.IsNullOrEmpty(ipEnd.Octet3) Then Return False
                    If String.IsNullOrEmpty(ipEnd.Octet4) Then Return False

                    Return True
                End Get
            End Property

#End Region

#Region "Private Fields"

            Private ReadOnly _import As Boolean
            Private _portScanner As Scanner
            Private _scanning As Boolean = False

#End Region

#Region "Private Methods"

#Region "Event Handlers"

            Private Sub PortScan_Load(sender As Object, e As EventArgs) Handles MyBase.Load
                ApplyLanguage()

                Try
                    If _import Then
                        lvHosts.Columns.AddRange(New ColumnHeader() _
                                                    {clmHost, clmSSH, clmTelnet, clmHTTP, clmHTTPS, clmRlogin, clmRDP,
                                                     clmVNC})
                        ShowImportControls(True)
                        cbProtocol.SelectedIndex = 0
                    Else
                        lvHosts.Columns.AddRange(New ColumnHeader() {clmHost, clmOpenPorts, clmClosedPorts})
                        ShowImportControls(False)
                    End If
                Catch ex As Exception
                    Runtime.MessageCollector.AddExceptionMessage(Language.Language.strPortScanCouldNotLoadPanel, ex)
                End Try
            End Sub

            Private Sub portStart_Enter(sender As Object, e As EventArgs) Handles portStart.Enter
                portStart.Select(0, portStart.Text.Length)
            End Sub

            Private Sub portEnd_Enter(sender As Object, e As EventArgs) Handles portEnd.Enter
                portEnd.Select(0, portEnd.Text.Length)
            End Sub

            Private Sub btnScan_Click(sender As Object, e As EventArgs) Handles btnScan.Click
                If _scanning Then
                    StopScan()
                Else
                    If IpsValid Then
                        StartScan()
                    Else
                        Runtime.MessageCollector.AddMessage(MessageClass.WarningMsg,
                                                            Language.Language.strCannotStartPortScan)
                    End If
                End If
            End Sub

            Private Sub btnImport_Click(sender As Object, e As EventArgs) Handles btnImport.Click
                Dim protocol As Protocols = Misc.StringToEnum(GetType(Protocols), cbProtocol.SelectedItem)

                Dim hosts As New List(Of ScanHost)
                For Each item As ListViewItem In lvHosts.SelectedItems
                    Dim scanHost = TryCast(item.Tag, ScanHost)
                    If scanHost IsNot Nothing Then hosts.Add(item.Tag)
                Next

                Import.ImportFromPortScan(hosts, protocol)

                DialogResult = DialogResult.OK
                Close()
            End Sub

#End Region

            Private Sub ApplyLanguage()
                lblStartIP.Text = String.Format("{0}:", Language.Language.strStartIP)
                lblEndIP.Text = String.Format("{0}:", Language.Language.strEndIP)
                btnScan.Text = Language.Language.strButtonScan
                btnImport.Text = Language.Language.strButtonImport
                lblOnlyImport.Text = String.Format("{0}:", Language.Language.strProtocolToImport)
                clmHost.Text = Language.Language.strColumnHostnameIP
                clmOpenPorts.Text = Language.Language.strOpenPorts
                clmClosedPorts.Text = Language.Language.strClosedPorts
                Label2.Text = String.Format("{0}:", Language.Language.strEndPort)
                Label1.Text = String.Format("{0}:", Language.Language.strStartPort)
                TabText = Language.Language.strMenuPortScan
                Text = Language.Language.strMenuPortScan
            End Sub

            Private Sub ShowImportControls(controlsVisible As Boolean)
                pnlPorts.Visible = controlsVisible
                pnlImport.Visible = controlsVisible
                If controlsVisible Then
                    lvHosts.Height = pnlImport.Top - lvHosts.Top
                Else
                    lvHosts.Height = pnlImport.Bottom - lvHosts.Top
                End If
            End Sub

            Private Sub StartScan()
                Try
                    _scanning = True
                    SwitchButtonText()
                    lvHosts.Items.Clear()

                    Dim ipAddressStart As IPAddress = IPAddress.Parse(ipStart.Text)
                    Dim ipAddressEnd As IPAddress = IPAddress.Parse(ipEnd.Text)

                    If _import Then
                        _portScanner = New Scanner(ipAddressStart, ipAddressEnd)
                    Else
                        _portScanner = New Scanner(ipAddressStart, ipAddressEnd, portStart.Value, portEnd.Value)
                    End If

                    AddHandler _portScanner.BeginHostScan, AddressOf PortScanner_BeginHostScan
                    AddHandler _portScanner.HostScanned, AddressOf PortScanner_HostScanned
                    AddHandler _portScanner.ScanComplete, AddressOf PortScanner_ScanComplete

                    _portScanner.StartScan()
                Catch ex As Exception
                    Runtime.MessageCollector.AddMessage(MessageClass.ErrorMsg,
                                                        "StartScan failed (UI.Window.PortScan)" & vbNewLine &
                                                        ex.ToString(), True)
                End Try
            End Sub

            Private Sub StopScan()
                If _portScanner IsNot Nothing Then _portScanner.StopScan()
                _scanning = False
                SwitchButtonText()
            End Sub

            Private Sub SwitchButtonText()
                If _scanning Then
                    btnScan.Text = Language.Language.strButtonStop
                Else
                    btnScan.Text = Language.Language.strButtonScan
                End If

                prgBar.Maximum = 100
                prgBar.Value = 0
            End Sub

            Private Shared Sub PortScanner_BeginHostScan(host As String)
                Runtime.MessageCollector.AddMessage(MessageClass.InformationMsg, "Scanning " & host, True)
            End Sub

            Private Delegate Sub PortScannerHostScannedDelegate _
                (host As ScanHost, scannedCount As Integer, totalCount As Integer)

            Private Sub PortScanner_HostScanned(host As ScanHost, scannedCount As Integer, totalCount As Integer)
                If InvokeRequired Then
                    Invoke(New PortScannerHostScannedDelegate(AddressOf PortScanner_HostScanned),
                           New Object() {host, scannedCount, totalCount})
                    Return
                End If

                Runtime.MessageCollector.AddMessage(MessageClass.InformationMsg, "Host scanned " & host.HostIp, True)

                Dim listViewItem As ListViewItem = host.ToListViewItem(_import)
                If listViewItem IsNot Nothing Then
                    lvHosts.Items.Add(listViewItem)
                    listViewItem.EnsureVisible()
                End If

                prgBar.Maximum = totalCount
                prgBar.Value = scannedCount
            End Sub

            Private Delegate Sub PortScannerScanComplete(hosts As List(Of ScanHost))

            Private Sub PortScanner_ScanComplete(hosts As List(Of ScanHost))
                If InvokeRequired Then
                    Invoke(New PortScannerScanComplete(AddressOf PortScanner_ScanComplete), New Object() {hosts})
                    Return
                End If

                Runtime.MessageCollector.AddMessage(MessageClass.InformationMsg, Language.Language.strPortScanComplete)

                _scanning = False
                SwitchButtonText()
            End Sub

#End Region
        End Class
    End Namespace

End Namespace

File: mRemoteV1/CS/UI/UI.Window.PortScan.cs
using System.Collections.Generic;
using System;
using AxWFICALib;
using System.Drawing;
using System.Diagnostics;
using System.Data;
using AxMSTSCLib;
using Microsoft.VisualBasic;
using System.Collections;
using System.Windows.Forms;
//using mRemoteNG.Tools.PortScan;
using mRemoteNG.My;
using WeifenLuo.WinFormsUI.Docking;
using mRemoteNG.App;


namespace mRemoteNG.UI.Window
{
	public partial class PortScan : Base
	{
        #region Constructors
		public PortScan(DockContent panel, bool import)
		{
			InitializeComponent();
					
			WindowType = Type.PortScan;
			DockPnl = panel;
			_import = import;
		}
        #endregion
				
        #region Private Properties
        private bool IpsValid
		{
			get
			{
				if (string.IsNullOrEmpty(ipStart.Octet1))
				{
					return false;
				}
				if (string.IsNullOrEmpty(ipStart.Octet2))
				{
					return false;
				}
				if (string.IsNullOrEmpty(ipStart.Octet3))
				{
					return false;
				}
				if (string.IsNullOrEmpty(ipStart.Octet4))
				{
					return false;
				}
						
				if (string.IsNullOrEmpty(ipEnd.Octet1))
				{
					return false;
				}
				if (string.IsNullOrEmpty(ipEnd.Octet2))
				{
					return false;
				}
				if (string.IsNullOrEmpty(ipEnd.Octet3))
				{
					return false;
				}
				if (string.IsNullOrEmpty(ipEnd.Octet4))
				{
					return false;
				}
						
				return true;
			}
		}
        #endregion
				
        #region Private Fields
		private bool _import;
		private Tools.PortScan.Scanner _portScanner;
		private bool _scanning = false;
        #endregion
				
        #region Private Methods
        #region Event Handlers
		public void PortScan_Load(System.Object sender, EventArgs e)
		{
			ApplyLanguage();
					
			try
			{
				if (_import)
				{
					lvHosts.Columns.AddRange(new ColumnHeader[] {clmHost, clmSSH, clmTelnet, clmHTTP, clmHTTPS, clmRlogin, clmRDP, clmVNC});
					ShowImportControls(true);
					cbProtocol.SelectedIndex = 0;
				}
				else
				{
					lvHosts.Columns.AddRange(new ColumnHeader[] {clmHost, clmOpenPorts, clmClosedPorts});
					ShowImportControls(false);
				}
			}
			catch (Exception ex)
			{
				Runtime.MessageCollector.AddExceptionMessage(My.Language.strPortScanCouldNotLoadPanel, ex);
			}
		}
				
		public void portStart_Enter(System.Object sender, EventArgs e)
		{
			portStart.Select(0, portStart.Text.Length);
		}
				
		public void portEnd_Enter(System.Object sender, EventArgs e)
		{
			portEnd.Select(0, portEnd.Text.Length);
		}
				
		public void btnScan_Click(System.Object sender, EventArgs e)
		{
			if (_scanning)
			{
				StopScan();
			}
			else
			{
				if (IpsValid)
				{
					StartScan();
				}
				else
				{
					Runtime.MessageCollector.AddMessage(Messages.MessageClass.WarningMsg, My.Language.strCannotStartPortScan);
				}
			}
		}
				
		public void btnImport_Click(System.Object sender, EventArgs e)
		{
			mRemoteNG.Connection.Protocol.Protocols protocol = Tools.Misc.StringToEnum(typeof(mRemoteNG.Connection.Protocol.Protocols), System.Convert.ToString(cbProtocol.SelectedItem));
					
			List<Tools.PortScan.ScanHost> hosts = new List<Tools.PortScan.ScanHost>();
			foreach (ListViewItem item in lvHosts.SelectedItems)
			{
				Tools.PortScan.ScanHost scanHost = item.Tag as Tools.PortScan.ScanHost;
				if (scanHost != null)
				{
					hosts.Add(item.Tag);
				}
			}
					
			App.Import.ImportFromPortScan(hosts, protocol);
					
			DialogResult = DialogResult.OK;
			Close();
		}
        #endregion
				
		private void ApplyLanguage()
		{
			lblStartIP.Text = string.Format("{0}:", My.Language.strStartIP);
			lblEndIP.Text = string.Format("{0}:", My.Language.strEndIP);
			btnScan.Text = My.Language.strButtonScan;
			btnImport.Text = My.Language.strButtonImport;
			lblOnlyImport.Text = string.Format("{0}:", My.Language.strProtocolToImport);
			clmHost.Text = My.Language.strColumnHostnameIP;
			clmOpenPorts.Text = My.Language.strOpenPorts;
			clmClosedPorts.Text = My.Language.strClosedPorts;
			Label2.Text = string.Format("{0}:", My.Language.strEndPort);
			Label1.Text = string.Format("{0}:", My.Language.strStartPort);
			TabText = My.Language.strMenuPortScan;
			Text = My.Language.strMenuPortScan;
		}
				
		private void ShowImportControls(bool controlsVisible)
		{
			pnlPorts.Visible = controlsVisible;
			pnlImport.Visible = controlsVisible;
			if (controlsVisible)
			{
				lvHosts.Height = pnlImport.Top - lvHosts.Top;
			}
			else
			{
				lvHosts.Height = pnlImport.Bottom - lvHosts.Top;
			}
		}
				
		private void StartScan()
		{
			try
			{
				_scanning = true;
				SwitchButtonText();
				lvHosts.Items.Clear();
						
				System.Net.IPAddress ipAddressStart = System.Net.IPAddress.Parse(ipStart.Text);
				System.Net.IPAddress ipAddressEnd = System.Net.IPAddress.Parse(ipEnd.Text);
						
				if (_import)
				{
					_portScanner = new Tools.PortScan.Scanner(ipAddressStart, ipAddressEnd);
				}
				else
				{
					_portScanner = new Tools.PortScan.Scanner(ipAddressStart, ipAddressEnd, (int) portStart.Value, (int) portEnd.Value);
				}
						
				_portScanner.BeginHostScan += PortScanner_BeginHostScan;
				_portScanner.HostScanned += PortScanner_HostScanned;
				_portScanner.ScanComplete += PortScanner_ScanComplete;
						
				_portScanner.StartScan();
			}
			catch (Exception ex)
			{
				Runtime.MessageCollector.AddMessage(Messages.MessageClass.ErrorMsg, "StartScan failed (UI.Window.PortScan)" + Constants.vbNewLine + ex.Message, true);
			}
		}
				
		private void StopScan()
		{
			if (_portScanner != null)
			{
				_portScanner.StopScan();
			}
			_scanning = false;
			SwitchButtonText();
		}
				
		private void SwitchButtonText()
		{
			if (_scanning)
			{
				btnScan.Text = My.Language.strButtonStop;
			}
			else
			{
				btnScan.Text = My.Language.strButtonScan;
			}
					
			prgBar.Maximum = 100;
			prgBar.Value = 0;
		}
				
		private static void PortScanner_BeginHostScan(string host)
		{
			Runtime.MessageCollector.AddMessage(Messages.MessageClass.InformationMsg, "Scanning " + host, true);
		}
				
		private delegate void PortScannerHostScannedDelegate(Tools.PortScan.ScanHost host, int scannedCount, int totalCount);
		private void PortScanner_HostScanned(Tools.PortScan.ScanHost host, int scannedCount, int totalCount)
		{
			if (InvokeRequired)
			{
				Invoke(new PortScannerHostScannedDelegate(PortScanner_HostScanned), new object[] {host, scannedCount, totalCount});
				return ;
			}
					
			Runtime.MessageCollector.AddMessage(Messages.MessageClass.InformationMsg, "Host scanned " + host.HostIp, true);
					
			ListViewItem listViewItem = host.ToListViewItem(_import);
			if (listViewItem != null)
			{
				lvHosts.Items.Add(listViewItem);
				listViewItem.EnsureVisible();
			}
					
			prgBar.Maximum = totalCount;
			prgBar.Value = scannedCount;
		}
				
		private delegate void PortScannerScanComplete(List<Tools.PortScan.ScanHost> hosts);
		private void PortScanner_ScanComplete(List<Tools.PortScan.ScanHost> hosts)
		{
			if (InvokeRequired)
			{
				Invoke(new PortScannerScanComplete(PortScanner_ScanComplete), new object[] {hosts});
				return ;
			}
					
			Runtime.MessageCollector.AddMessage(Messages.MessageClass.InformationMsg, Language.strPortScanComplete);
					
			_scanning = false;
			SwitchButtonText();
		}
        #endregion
	}
}
